format/yaml/src/main/java/org/spongepowered/configurate/yaml/YamlConstructor.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Configurate
+ * Copyright (C) zml and Configurate contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.spongepowered.configurate.yaml;
+
+import org.checkerframework.checker.nullness.qual.EnsuresNonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.spongepowered.configurate.CommentedConfigurationNode;
+import org.spongepowered.configurate.ConfigurationNode;
+import org.spongepowered.configurate.ConfigurationOptions;
+import org.yaml.snakeyaml.LoaderOptions;
+import org.yaml.snakeyaml.comments.CommentLine;
+import org.yaml.snakeyaml.constructor.Constructor;
+import org.yaml.snakeyaml.nodes.MappingNode;
+import org.yaml.snakeyaml.nodes.Node;
+import org.yaml.snakeyaml.nodes.NodeId;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+class YamlConstructor extends Constructor {
+
+    private static final Pattern LINE_BREAK_PATTERN = Pattern.compile("\\R");
+
+    @Nullable ConfigurationOptions options;
+
+    YamlConstructor(final LoaderOptions loadingConfig) {
+        super(loadingConfig);
+    }
+
+    @Override
+    @EnsuresNonNull("options")
+    public Object getSingleData(final Class<?> type) {
+        if (this.options == null) {
+            throw new IllegalStateException("options must be set before calling load!");
+        }
+        return super.getSingleData(type);
+    }
+
+    @Override
+    protected Object constructObjectNoCheck(final Node yamlNode) {
+        //noinspection DataFlowIssue guarenteed NonNull by getSingleData, which load(Reader) uses
+        final CommentedConfigurationNode node = CommentedConfigurationNode.root(this.options);
```

Comment:
I'm really not a fan of the fact that by creating a new root node here, we effectively construct two nodes per node when loading in a file

Reply:
You're then mainly talking about the mapping part right? When looking at the differences between a `root` node and a `.node` node it looks like only the path and parent differs.
We could add a method that works essentially like `attachIfNecessary` / `parentEnsureAttached`, but allows you to set the parent & path as well.